src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java

```diff
@@ -305,19 +308,40 @@ static void assertNoOverlap(InitCodeNode root, List<String> paths) {
     // that reference the same nodes.
     HashMap<InitCodeNode, String> refFrom = new HashMap<>();
 
+    // Keep track of the resource name entities. Configuring an entity twice or configuring an
+    // entity
+    // and the parent node at the same time will cause collision. Configuring two different entities
+    // will not.
+    Multimap<InitCodeNode, String> nodeEntities = HashMultimap.create();
+
     // Below we'll perform depth-first search, keep a list of nodes we've seen but have not
     // descended into. It doesn't really matter if we search breath- or depth-first; DFS is a little
     // more efficient on average.
     ArrayDeque<InitCodeNode> subNodes = new ArrayDeque<>();
 
     for (String path : paths) {
       subNodes.add(root.subTree(path));
+      String entity = FieldStructureParser.parseEntityName(path);
       while (!subNodes.isEmpty()) {
         InitCodeNode node = subNodes.pollLast();
         String oldPath = refFrom.put(node, path);
-        if (oldPath != null) {
-          throw new IllegalArgumentException(
-              String.format("SampleInitAttribute %s overlaps with %s", oldPath, path));
+        if (oldPath == null) {
+          checkArgument(
+              entity == null || nodeEntities.put(node, entity),
+              "SampleInitAttribute %s overlaps with %s",
+              path,
+              path);
+        } else {
+          checkArgument(
+              entity != null && nodeEntities.containsKey(node),
```

Comment:
Am I  understanding this correctly?: Because oldPath was not null, the node was previously inserted. Because the node is being referenced again, it must be because it has an entity. And if it has an entity now, it means that the previous reference must have referenced an entity. The assertion checks these two "must"s.

If this is right, then the error message is not helpful. Also, adding in a comment the brief explanation I put above would help me in the future (and others not deeply immersed in this part of the code to begin with)